src/main/kotlin/org/rust/lang/core/types/Substitution.kt

```diff
@@ -0,0 +1,66 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.lang.core.types
+
+import org.rust.lang.core.psi.RsLifetimeParameter
+import org.rust.lang.core.psi.RsTypeParameter
+import org.rust.lang.core.types.infer.TypeFolder
+import org.rust.lang.core.types.infer.substitute
+import org.rust.lang.core.types.re.Re
+import org.rust.lang.core.types.re.ReEarlyBound
+import org.rust.lang.core.types.ty.Ty
+import org.rust.lang.core.types.ty.TyTypeParameter
+import org.rust.lang.core.types.ty.TyUnknown
+import org.rust.stdext.zipValues
+
+open class Substitution(
+    val typeSubst: Map<TyTypeParameter, Ty> = emptyMap(),
+    val lifetimeSubst: Map<ReEarlyBound, Re> = emptyMap()
+) {
+    val types: Collection<Ty> get() = typeSubst.values
+    val lifetimes: Collection<Re> get() = lifetimeSubst.values
+    val kinds: Collection<Kind> get() = types + lifetimes
+
+    operator fun plus(other: Substitution): Substitution =
+        Substitution(
+            typeSubst + other.typeSubst,
```

Comment:
Since now we have a single place of maps merging, use `HashMap`s here instead of `LinkedHashMap`, produced by kotlin's `+`. It may looks like `HashMap(typeSubst).apply { putAll(other.typeSubst) }`. But really we can write our own map merging function with some optimizations, i.e. if both maps are empty, return `emptyMap()`. I didn't do it previously because of multiple points of merging, but now it's the single place. Mmm, anyway, we can do in a separate PR....